consul_exporter.go

```diff
@@ -334,6 +352,11 @@ func main() {
 	kingpin.Flag("consul.server-name", "When provided, this overrides the hostname for the TLS certificate. It can be used to ensure that the certificate name matches the hostname we declare.").Default("").StringVar(&opts.serverName)
 	kingpin.Flag("consul.timeout", "Timeout on HTTP requests to consul.").Default("200ms").DurationVar(&opts.timeout)
 
+	// Query options.
+	kingpin.Flag("consul.allow_stale", "Allows any Consul server (non-leader) to service a read.").Default("true").BoolVar(&queryOptions.AllowStale)
+	kingpin.Flag("consul.require_consistent", "Forces the read to be fully consistent.").Default("false").BoolVar(&queryOptions.RequireConsistent)
+	kingpin.Flag("consul.datacenter", "Overwrites the DC provided by the Config.").Default("").StringVar(&queryOptions.Datacenter)
```

Comment:
What's the use case for that? The exporter should be deployed in the same datacenter.

Reply:
For example we have a number of datacenters with consul, but one of them doesn't have marathon, which means that we will want to run a  consul_exporter for this datacenter from another datacenter.

Reply:
But then you can also just change the consul address flag. I'm not I see the need to add a second way to solve the problem.

Reply:
* Using the consul address flag you can query a remote agent for its local dc
* Using the dc flag you can query a local agent for a remote datacenter

If you don't require consistent reads and allow stale reads I suspect that querying the local agent will be way faster, in particular when you remote datacenter is a few continents away.

Reply:
Unless anything has changed in consul since version 0.5/0.6 to that regard, the local consul agent will simply forward the HTTP request to an agent in the remote datacenter. The local cluster won't cache the data of a remote datacenter (please prove me wrong with a link to docs/code otherwise) So this essentially introduces another network hop and another point of failure to retrieve stats. The latter goes against the Prometheus philosophy to run exporters as close as possible to the target, so I'm still not convinced we should add a second way to that.

Reply:
As explained above, we have uses cases where the exporter needs to run in another datacenter because there is no infrastructure to run random programs there. In that case it seems more logical to get the information from the local agent that to try to point to a random agent somewhere in the remote datacenter (or to point to the master, directly).

Without that, I would have to rely on consul DNS to find a working consul agent in the remote datacenter, which seems more brittle that using the local agent.

Reply:
ping ?

Reply:
It's not a common nor recommended setup to scrape a consul cluster from a consul_exporter in a different datacenter. Having this flag makes it easy for users to follow such a anti-pattern in Prometheus exporter configuration. Exporters should run as close as possible to the source of the data, which means on a consul server node in this case.

I'm happy to merge a PR without this datacenter flag and continue the discussion around this flag in a separate PR.

Reply:
In our case, we have two separate Consul clusters and "datacenters" in the same physical datacenter, one for our production environment and one for our QA environment.  We would like one Prometheus instance, in the same physical datacenter as these two Consul clusters to ingest metrics from both of these consul clusters.

According to the answer in this question about the best way to run this exporter, we should ideally be running this exporter next to the Prometheus process and pointing it to a local Consul agent:

https://github.com/prometheus/consul_exporter/issues/42

These Consul clusters are federated, so the local Consul agent knows about both Consul clusters, but there is no way to get metrics from one of them with this exporter.

Like @iksaif says earlier, I guess I can pick a random Consul agent from each cluster, run an exporter on it, and scrape it, but that seems worse than having an option in this exporter to specify a datacenter.